### ISSUE TITLE
.refine does not work in request params and request query

### ISSUE THREAD

**Issue by ootkin** [asteasolutions/zod-to-openapi#198]
If we use `refine` to add complex validation of a zod schema, the output is a ZodEffect, and it does not work with `request.query` and `request.params`. It works with `body` tho.

I've got this schema:
```ts
  const user = z
    .object({ name: z.string(), age: z.string() })
    .refine((val) => val.name || val.age, {
      message: "name or age required",
      path: ["name", "age"],
    });
```

and this path:
```ts
      registry.registerPath({
        method: "get",
        path: "/users/{id}",
        summary: "Get a single user",
        request: {
          params: user, // <-- error
          query: user, // <-- error
          body: {
            content: {
              'application/json': {
                schema: user
              }
            }
          }
        },
        responses: {},
      });
```

The error is:
```
Type 'ZodEffects<ZodObject<{ age: ZodString; name3: ZodString; }, "strip", ZodTypeAny, { name3: string; age: string; }, { name3: string; age: string; }>, { ...; }, { ...; }>' is not assignable to type 'ZodObject<any, any, any, { [x: string]: any; }, { [x: string]: any; }>'
```

In the source code, the schema body type is `ZodType<unknown> | SchemaObject | ReferenceObject`
While params and query are: `AnyZodObject` which is `ZodObject<any, any, any>`

**Comment by AGalabov** [comment]
Hello again @ootkin! 

You are correct. Overall the idea that we've followed is that `query`, `params` and `headers` are not just any type - they are always an object (or even a key-value entries) because of how HTTP specifications are being followed. And in the OpenAPI realm they are described in a different manner than the body. We chose to use the AnyZodObject instead of a regular `Record` type from typescript mainly for easier usage. The rest of the types (like just a number) are not compatible with the OpenAPI specification.

Now I do imagine your question would be - sure but can you not make it so that `ZodEffect`s with a `ZodObject` base can be used - to which the answer might be "Yes" and I'll have to check.

In general our main logic here was that on an API docs level the documentation is generated based on "an object of properties" => the documentation part can be rendered as such. And any additional validation can easily be done on another layer of an API as it contains a somewhat business specific logic.

i.e you could always do:
```ts
  const user = z.object({ name: z.string(), age: z.string() })

  const organizationUser = user
    .refine((val) => val.name || val.age, {
      message: "name or age required",
      path: ["name", "age"],
    });
```
and use them accordingly. However I do agree that this might not be optimal so I'd try and look into the possibility of using refine/preprocess on a `ZodObject`

**Comment by thelinuxlich** [comment]
The generator should just unwrap when detecting a zodeffect

**Comment by AGalabov** [comment]
@thelinuxlich  you might be correct - if you've checked the code - please go ahead and open a PR

**Comment by AGalabov** [comment]
@ootkin @thelinuxlich this should now be available as part of the [v7.0.0 Release](https://github.com/asteasolutions/zod-to-openapi/releases/tag/v7.0.0) :rocket: .Sorry that this took so long
